parity-crypto/src/error.rs

```diff
@@ -18,80 +18,113 @@ use rscrypt;
 use block_modes;
 use aes_ctr;
 use std::error::Error as StdError;
+use std::{fmt, result};
 
-quick_error! {
-	#[derive(Debug)]
-	pub enum Error {
-		Scrypt(e: ScryptError) {
-			cause(e)
-			from()
-		}
-		Symm(e: SymmError) {
-			cause(e)
-			from()
-		}
-		AsymShort(det: &'static str) {
-			description(det)
-		}
-		AsymFull(e: Box<dyn StdError + Send>) {
-			cause(&**e)
-			description(e.description())
+#[derive(Debug)]
+pub enum Error {
+	Scrypt(ScryptError),
+	Symm(SymmError),
+	// TODO: used anywhere?
+	AsymShort(&'static str),
+	// TODO: used anywhere?
+	AsymFull(Box<dyn StdError + Send>)
+}
+
+#[derive(Debug)]
+pub enum ScryptError {
+	// log(N) < r / 16
+	InvalidN,
+	// p <= (2^31-1 * 32)/(128 * r)
+	InvalidP,
+	ScryptParam(rscrypt::errors::InvalidParams),
+	ScryptLength(rscrypt::errors::InvalidOutputLen),
+}
+
+#[derive(Debug)]
+pub struct SymmError(PrivSymmErr);
+
+#[derive(Debug)]
+enum PrivSymmErr {
+	Offset(usize), // TODO: never constructed – can remove?
+	BlockMode(block_modes::BlockModeError),
+	KeyStream(aes_ctr::stream_cipher::LoopError),
+	InvalidKeyLength(block_modes::InvalidKeyIvLength),
+}
+
+impl StdError for Error {
+	fn source(&self) -> Option<&(StdError + 'static)> {
+		match self {
+			Error::Scrypt(scrypt_err) => Some(scrypt_err),
+			Error::Symm(symm_err) => Some(symm_err),
+			Error::AsymShort(_)=> None,
+			Error::AsymFull(err)=> Some(&**err),
 		}
 	}
 }
 
-impl Into<std::io::Error> for Error {
-	fn into(self) -> std::io::Error {
-		std::io::Error::new(std::io::ErrorKind::Other, format!("Crypto error: {}",self))
+impl StdError for ScryptError {
+	fn source(&self) -> Option<&(StdError + 'static)> {
+		match self {
+			ScryptError::ScryptParam(err) => Some(err),
+			ScryptError::ScryptLength(err) => Some(err),
+			_ => None,
+		}
 	}
 }
 
-quick_error! {
-	#[derive(Debug)]
-	pub enum ScryptError {
-		// log(N) < r / 16
-		InvalidN {
-			display("Invalid N argument of the scrypt encryption")
-		}
-		// p <= (2^31-1 * 32)/(128 * r)
-		InvalidP {
-			display("Invalid p argument of the scrypt encryption")
-		}
-		ScryptParam(e: rscrypt::errors::InvalidParams) {
-			display("invalid params for scrypt: {}", e)
-			cause(e)
-			from()
-		}
-		ScryptLength(e: rscrypt::errors::InvalidOutputLen) {
-			display("invalid scrypt output length: {}", e)
-			cause(e)
-			from()
+impl StdError for SymmError {
+	fn source(&self) -> Option<&(StdError + 'static)> {
+		match &self.0 {
+			PrivSymmErr::BlockMode(err) => Some(err),
+			// TODO: the trait `std::error::Error` is not implemented for
```

Comment:
hm, looks like this is only implemented with `std` feature enable and  `aes-ctr` [doesn't](https://github.com/RustCrypto/stream-ciphers/blob/d59145d8bb100bf077274bc5e5ae8155f2472e35/aes-ctr/Cargo.toml#L13) use it